src/compiler/program.ts

```diff
@@ -129,86 +129,65 @@ namespace ts {
         skipTsx: boolean;
     }
 
-    function tryReadTypesSection(packageJsonPath: string, baseDirectory: string, state: ModuleResolutionState): string {
-        let jsonContent: { typings?: string, types?: string };
-        try {
-            const jsonText = state.host.readFile(packageJsonPath);
-            jsonContent = jsonText ? <{ typings?: string, types?: string }>JSON.parse(jsonText) : {};
-        }
-        catch (e) {
-            // gracefully handle if readFile fails or returns not JSON
-            jsonContent = {};
-        }
-
-        let typesFile: string;
-        let fieldName: string;
-        // first try to read content of 'typings' section (backward compatibility)
-        if (jsonContent.typings) {
-            if (typeof jsonContent.typings === "string") {
-                fieldName = "typings";
-                typesFile = jsonContent.typings;
+    function tryFetchField(content: any, section: string, state: ModuleResolutionState): string {
+        const field = content[section];
+        if (field) {
+            if (typeof field === "string") {
+                return field;
             }
             else {
                 if (state.traceEnabled) {
-                    trace(state.host, Diagnostics.Expected_type_of_0_field_in_package_json_to_be_string_got_1, "typings", typeof jsonContent.typings);
+                    trace(state.host, Diagnostics.Expected_type_of_0_field_in_package_json_to_be_string_got_1, section, typeof field);
                 }
             }
         }
-        // then read 'types'
-        if (!typesFile && jsonContent.types) {
-            if (typeof jsonContent.types === "string") {
-                fieldName = "types";
-                typesFile = jsonContent.types;
-            }
-            else {
-                if (state.traceEnabled) {
-                    trace(state.host, Diagnostics.Expected_type_of_0_field_in_package_json_to_be_string_got_1, "types", typeof jsonContent.types);
-                }
-            }
-        }
-        if (typesFile) {
-            const typesFilePath = normalizePath(combinePaths(baseDirectory, typesFile));
-            if (state.traceEnabled) {
-                trace(state.host, Diagnostics.package_json_has_0_field_1_that_references_2, fieldName, typesFile, typesFilePath);
-            }
-            return typesFilePath;
-        }
-        return undefined;
     }
 
-   function tryReadMainSection(packageJsonPath: string, baseDirectory: string, state: ModuleResolutionState): string {
-        let jsonContent: { main?: string };
+    function tryReadPathFromConfigSection(packageJsonPath: string, sections: string | string[], baseDirectory: string, state: ModuleResolutionState): string {

```

Comment:
It seems like this could be even simpler:

```ts
function getPackageEntry(packageJson: any, key: string, tag: string, state: ModuleResolutionState) {
    const value = packageJson[key];
    if (typeof value === tag) {
        return value;
    }
    if (state.traceEnabled) {
        trace(state.host, Diagnostics.Expected_type_of_0_field_in_package_json_to_be_1_got_2, key, tag, typeof value);
    }
    return undefined;
}

function getPackageEntryAsPath(packageJson: any, key: string, state: ModuleResolutionState) {
    const value = getPackageEntry(packageJson, key, "string", state);
    const path = value ? normalizePath(combinePaths(getDirectoryPath(packageJson), value)) : undefined;
    if (path && state.traceEnabled) {
        trace(state.host, Diagnostics.package_json_has_0_field_1_that_references_2, key, value, path);
    }
    return path;
}

function getPackageTypes(packageJsonPath: string, state: ModuleResolutionState) {
    const { config } = readConfigFile(packageJsonPath, state.host.readFile);
    if (config) {
        return getPackageEntryAsPath(config, "typings", state)
            || getPackageEntryAsPath(config, "types", state);
    }
    return undefined;
}

function getPackageMain(packageJsonPath: string, state: ModuleResolutionState) {
    const { config } = readConfigFile(packageJsonPath, state.host.readFile);
    if (config) {
        return getPackageEntryAsPath(packageJson, "main", state);
    }
    return undefined;
}
```